File: routes/recipes.js
var express = require("express");
var router = express.Router();

var assert = require("assert");
var fetch = require("node-fetch");

var MongoClient = require("mongodb").MongoClient;
var db_conn_str = process.env.DB_CONN_STR;
if (!db_conn_str) {
  console.error("DB_CONN_STR variable not set!");
}

const APIurl = "https://api.recipe.studio/";

// GET: default
router.get("/", (req, res, next) => {
  res.status(400).json({
    message: "No recipe id was sent. Use /recipe/all for all recipes"
  });
});

// GET: get all recipes
router.get("/all", mongoGetAll);

// GET: get recipes from search
router.get("/search", mongoSearch);

// GET: get recipe by :id
router.get("/:id", mongoGet);

// POST: create recipe
router.post("/new", mongoPost);

// PUT: update recipe by :id
router.put("/:id", mongoPut);

// DELETE: delete recipe by :id
router.delete("/:id", mongoDel);

// DB read one function
function mongoGet(req, res, next) {
  let mongodb = require("mongodb");

  // check for correct ID length
  let rid = req.params.id;
  if (rid.length != 24) {
    res.status(400).json({ message: "Invalid recipe ID" });
    return;
  }

  // connect to db
  try {
    MongoClient.connect(db_conn_str, (err, db) => {
      if (err) {
        console.error(err);
        res.status(500).json({ message: "Could not connect to database" });
      } else {
        db
          .db("recipe-studio")
          .collection("recipes")
          .findOne({ _id: new mongodb.ObjectID(rid) }, (err, recipe) => {
            // send response data
            if (!recipe) {
              res.status(404).json({ message: "Recipe " + rid + " not found" });
            } else {
              res.status(200).json(recipe);
            }
          });
      }

      db.close();
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Error fetching recipes", error: err });
  }
}

// DB read all function
function mongoGetAll(req, res, next) {
  // connect to db
  try {
    MongoClient.connect(db_conn_str, (err, db) => {
      assert.equal(null, err, "Could not connect to db");

      db
        .db("recipe-studio")
        .collection("recipes")
        .find({})
        .toArray((err, recipes) => {
          res.status(200).json(recipes);
        });

      db.close();
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Error fetching recipes", error: err });
  }
}

function mongoSearch(req, res, next) {
  // check if getting user recipes
  let searchObj = {};
  if (req.query.user) {
    searchObj["author.uid"] = req.query.user;
  }

  if (req.query.servings) {
    searchObj.servings = req.query.servings;
  }

  if (req.query.ingredients) {
    let ingQuery = req.query.ingredients;
    ingQuery = ingQuery.split(",");

    searchObj["ingredients.data._id"] = { $all: ingQuery };
  }

  // connect to db
  try {
    MongoClient.connect(db_conn_str, (err, db) => {
      if (err) {
        console.error(err);
        res
          .status(500)
          .json({ message: "Error connecting to database", error: err });
      } else {
        db
          .db("recipe-studio")
          .collection("recipes")
          .find(searchObj)
          .toArray((err, recipes) => {
            if (err) {
              console.error(err);
              res.status(500).json({ error: err });
            }
            res.status(200).json(recipes);
          });

        db.close();
      }
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Error getting recipes", error: err });
  }
}

// DB delete function
function mongoDel(req, res, next) {
  let mongodb = require("mongodb");

  // check for correct ID length
  let rid = req.params.id;
  if (rid.length != 24) {
    res.status(400).json({ message: "Invalid recipe ID" });
    return;
  }

  // connect to db
  try {
    MongoClient.connect(db_conn_str, (err, db) => {
      assert.equal(null, err, "Could not connect to db");

      db
        .db("recipe-studio")
        .collection("recipes")
        .deleteOne({ _id: new mongodb.ObjectID(rid) }, (err, r) => {
          assert.equal(null, err, "Could not delete recipe: " + err);

          res.status(200).json({ message: "Deleted recipe" });
        });

      db.close();
    });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: "Error deleting recipe", error: error });
  }
}

// DB create function
function mongoPost(req, res, next) {
  let d = new Date();

  formObject(req).then(newObj => {
    // connect to DB here
    try {
      MongoClient.connect(db_conn_str, (err, db) => {
        if (err) {
          console.error(err);
          res
            .status(500)
            .json({ message: "Could not connect to database", error: err });
        }

        // insert into db
        db
          .db("recipe-studio")
          .collection("recipes")
          .insertOne(newObj, (err, result) => {
            // send response data
            if (err) {
              console.error(err);
              res
                .status(500)
                .json({ message: "Error fetching data", error: err });
            } else if (result.insertedCount != 1) {
              res.status(500).json({ message: "Could not create recipe" });
            } else {
              res.status(200).json(newObj);
            }
          });
        db.close();
      });
    } catch (err) {
      console.error(err);
      res.status(500).json({ message: "Error creating recipe", error: err });
    }
  });
}

// DB update function
function mongoPut(req, res, next) {
  let mongodb = require("mongodb");

  // check for correct ID length
  let rid = req.params.id;
  if (rid.length != 24) {
    res.status(400).json({ message: "Invalid recipe ID" });
    return;
  }

  formObject(req).then(updateObj => {
    // connect to db
    try {
      MongoClient.connect(db_conn_str, (err, db) => {
        db
          .db("recipe-studio")
          .collection("recipes")
          .updateOne(
            { _id: new mongodb.ObjectID(rid) },
            { $set: updateObj },
            (err, r) => {
              if (err) {
                console.error(err);
                res
                  .status(500)
                  .json({ message: "Error while updating", error: err });
              } else {
                fetch(APIurl + "recipe/" + rid)
                  .then(res => {
                    return res.json();
                  })
                  .then(data => {
                    res.status(200).json(data);
                  });
              }
            }
          );
        db.close();
      });
    } catch (err) {
      console.error(err);
      res.status(500).json({ message: "Could not update recipe", error: err });
    }
  });
}

function formObject(req) {
  let objFields = {};
  if (req.body.name) {
    objFields.name = req.body.name;
  }
  if (req.body.directions) {
    objFields.directions = req.body.directions;
  }
  if (req.body.description) {
    objFields.description = req.body.description;
  }
  if (req.method == "POST") {
    // set initial image
    objFields.image = !req.body.image
      ? "http://via.placeholder.com/600x400"
      : req.body.image;

    // Promise for new object
    return new Promise((resolve, reject) => {
      // set date
      let d = new Date();
      objFields.created = d;

      // set user
      let a = validateAuthor(req.body.author).then(author => {
        objFields.author = author;
      });

      // set ingredients
      let i = validateIngredients(req.body.ingredients).then(ingredients => {
        objFields.ingredients = ingredients;
      });

      Promise.all([a, i]).then(() => {
        // return object
        resolve(objFields);
        reject(objFields);
      });
    });

    // Promise for update object
  } else {
    if (req.body.ingredients) {
      return new Promise((resolve, reject) => {
        validateIngredients(req.body.ingredients)
          .then(ingredients => {
            objFields.ingredients = ingredients;
          })
          .then(() => {
            resolve(objFields);
            reject({});
          });
      });
    } else {
      return new Promise((resolve, reject) => {
        resolve(objFields);
        reject(objFields);
      });
    }
  }
}

// Check author against database
function validateAuthor(uid) {
  let author = {};

  return new Promise((resolve, reject) => {
    fetch(APIurl + "user/" + uid)
      .then(res => {
        return res.json();
      })
      .then(author_data => {
        resolve({
          uid: author_data.uid,
          displayName: author_data.displayName
        });
        reject({
          uid: uid,
          displayName: "Unknown"
        });
      });
  });
}

// Check ingredients against database
function validateIngredients(ingredients_array) {
  let ingredients = [];
  let requests = [];

  // for each ingredient, create a promise to get ingredient data and
  //   push new data to array of objects with embedded old data.
  ingredients_array.forEach(i => {
    let promise = fetch(APIurl + "ingredient/" + i.ingredient)
      .then(res => {
        return res.json();
      })
      .then(ingredient_data => {
        ingredients.push({
          data: ingredient_data,
          quantity: i.quantity,
          units: i.units
        });
      });

    // push promise to array of promises
    requests.push(promise);
  });

  return new Promise((resolve, reject) => {
    // execute all promises
    Promise.all(requests).then(() => {
      resolve(ingredients);
      reject({});
    });
  });
}

module.exports = router;

File: routes/recipes.py
from bson import objectid, json_util
from flask import Blueprint, make_response, request
from pprint import pprint
from pymongo import MongoClient

import datetime
import json
import os


class Recipe:
    """
        Class to represent recipe objects locally.

        Attributes:
            r_id (str): Recipe ID (Mongo ObjectID)
            author (dict): Author data (uid, displayName)
            created (datetime): Date/time data in datetime format
            description (str): Recipe description
            directions (str): Recipe directions
            image (str): Image URL for recipe
            ingredients (Ingredient): Array of ingredient objects (see Ingredient class)
            name (str): Recipe name
    """

    r_id = ''
    author = {}
    created = ''
    description = ''
    directions = ''
    image = ''
    ingredients = []
    name = ''

    def __init__(self, author, created, description, directions, image, ingredients, name):
        self.author = author
        self.created = created
        self.description = description
        self.directions = directions
        self.image = image
        self.name = name

        # TODO: add Ingredient object parsing here.
        self.ingredients = ingredients

    def to_dict(self):
        _dict = {}
        _dict['_id'] = str(self.r_id)
        _dict['created'] = str(self.created)
        _dict['description'] = self.description
        _dict['directions'] = self.directions
        _dict['image'] = self.image
        _dict['name'] = self.name

        _dict['author'] = {}
        for key, val in self.author.items():
            _dict['author'][key] = val

        _dict['ingredients'] = []
        for ingredient in self.ingredients:
            _dict['ingredients'].append(ingredient)

        return _dict


class DbConnError(Exception):
    """Handler for errors in the database connection"""

    def __init__(self, message):
        self.message = message


# setup mongo client
# connect to mongo and get collection
DB_CONN_STR = os.getenv('DB_CONN_STR')
DB_NAME = os.getenv('DB_NAME')
RECIPE_COL = os.getenv('RECIPE_COL_NAME')

if (DB_CONN_STR is None):
    raise DbConnError('Variable DB_CONN_STR is not set')
if (DB_NAME is None):
    raise DbConnError('Variable DB_NAME is not set')
if (RECIPE_COL is None):
    raise DbConnError('Variable RECIPE_COL is not set')

_mongoclient = MongoClient(DB_CONN_STR)
_DB = _mongoclient[DB_NAME]
_RECIPES = _DB[RECIPE_COL]

# setup route blueprint
recipe = Blueprint('recipe_route', __name__)


@recipe.route('/')
def default():
    return make_response('Malformed request', 400)

# Fetch, update, delete recipe
@recipe.route('/<recipe_id>', methods=['GET', 'PUT', 'DELETE'])
def manage_recipe(recipe_id):

    response = {}
    if request.method == 'GET':
        response = get_one_recipe(recipe_id)
    elif request.method == 'PUT':
        response = update_recipe(recipe_id)
    elif request.method == 'DELETE':
        response = delete_recipe(recipe_id)
    else:
        response = make_response('Malformed request', 400)

    response.headers.add('Access-Control-Allow-Origin', '*')
    response.headers['Content-Type'] = 'application/json'
    return response


# Get all recipes
@recipe.route('/all', methods=['GET'])
def get_all_recipes():

    results = []
    raw_results = _RECIPES.find()
    for result in raw_results:
        formatted_result = read_bson(result)
        results.append(formatted_result)

    response = {}
    if result is not None:
        response = make_response(jsonify(results), 200)
    else:
        response = make_response(jsonify({}), 404)

    response.headers.add('Access-Control-Allow-Origin', '*')
    response.headers['Content-Type'] = 'application/json'
    return response

# Search recipes
@recipe.route('/search', methods=['GET'])
def search():
    data = '{}'
    response = make_response(jsonify(data), 200)
    response.headers.add('Access-Control-Allow-Origin', '*')
    response.headers['Content-Type'] = 'application/json'
    return response

# Create recipe
@recipe.route('/new', methods=['POST'])
def create_recipe():
    data = '{}'
    response = make_response(jsonify(data), 200)
    response.headers.add('Access-Control-Allow-Origin', '*')
    response.headers['Content-Type'] = 'application/json'
    return response


def get_one_recipe(r_id):
    result = _RECIPES.find_one({'_id': objectid.ObjectId(r_id)})
    if result is not None:
        formatted_result = read_bson(result)
        response = make_response(jsonify(formatted_result), 200)
    else:
        response = make_response(
            jsonify({
                'message': 'Recipe ' + r_id + ' not found'
            }), 404
        )

    return response


def update_recipe(r_id):
    return 'updated {}'.format(r_id)


def delete_recipe(r_id):
    return 'deleted {}'.format(r_id)


def read_bson(bson_data):
    # print('BSON')
    # pprint(bson_data)

    # extract data from bson and import to local Recipe object
    author = bson_data['author']
    created = bson_data['created']
    description = bson_data['description']
    directions = bson_data['directions']
    name = bson_data['name']
    image = bson_data['image']
    r_id = bson_data['_id']

    ingredients = []
    for ingredient in bson_data['ingredients']:
        ingredients.append(ingredient)

    recipe_obj = Recipe(
        author, created, description, directions, image, ingredients, name)

    recipe_obj.r_id = r_id
    recipe_dict = recipe_obj.to_dict()

    # print('JSON')
    # pprint(recipe_dict)
    return recipe_dict


def jsonify(obj):
    return json.dumps(obj, default=json_util.default)
